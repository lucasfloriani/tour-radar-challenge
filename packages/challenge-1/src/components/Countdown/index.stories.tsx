import React from "react";

import Countdown from "./index";

export default {
  title: "Design System/Countdown",
  component: Countdown,
};

const timeNow = new Date();
const fastEndTime = new Date(timeNow.getTime() + 1000 * 10);
const longEndTime = new Date(timeNow.getTime() + 1000 * 60 * 60 * 24 * 3 + 1000 * 60 * 30);

export const Default = () => <Countdown start={timeNow} end={longEndTime} />;

export const TimeEnding = () => <Countdown start={timeNow} end={fastEndTime} />;
