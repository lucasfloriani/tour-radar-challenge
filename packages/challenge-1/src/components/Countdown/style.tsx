import styled from "styled-components";
import Typography from "@tour-radar/design-system/components/Typography";

export const Wrapper = styled.section`
  display: grid;
  gap: 3px;
`;

type WrapperProps = {
  boxQty: number;
};

export const BoxWrapper = styled.div.attrs((props: WrapperProps) => ({
  boxQty: props?.boxQty,
}))`
  display: inline-grid;
  gap: 8px;
  grid-template-columns: ${({ boxQty }) => `repeat(${boxQty}, auto)`};
`;

export const CountdownTitle = styled(Typography).attrs({
  as: "h4",
  size: "small",
  weight: "600",
})`
  letter-spacing: 3px;
  text-transform: uppercase;
`;
