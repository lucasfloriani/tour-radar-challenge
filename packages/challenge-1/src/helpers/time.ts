export const SECOND = 1000;
export const MINUTE = SECOND * 60;
export const HOUR = MINUTE * 60;
export const DAY = HOUR * 24;

// * Helper functions and constants created so we don't need
// * to use a date library like date - fns(to reduce bundle size)
export const getDays = (time: number) => Math.floor(time / DAY);
export const getHours = (time: number) => Math.floor((time % DAY) / HOUR);
export const getMinutes = (time: number) => Math.floor((time % HOUR) / MINUTE);
export const getSeconds = (time: number) => Math.floor((time % MINUTE) / SECOND);

export const getTimeDiff = (startDate: Date, endDate: Date) => endDate.getTime() - startDate.getTime();

export const getTimeDiffObj = (startDate: Date, endDate: Date) => {
  const diff = getTimeDiff(startDate, endDate);
  return {
    days: getDays(diff),
    hours: getHours(diff),
    minutes: getMinutes(diff),
    seconds: getSeconds(diff),
  };
};
