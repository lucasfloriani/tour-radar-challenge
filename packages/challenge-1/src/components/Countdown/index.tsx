import { useEffect, useMemo, useState } from "react";
import { useTranslation } from "react-i18next";
import TimeBox from "@tour-radar/design-system/components/TimeBox";
import { getTimeDiffObj } from "helpers/time";
import { BoxWrapper, CountdownTitle, Wrapper } from "./style";

type CountdownBoxOptions = {
  seconds: boolean;
  minutes: boolean;
  hours: boolean;
  days: boolean;
};

type CountdownProps = {
  start?: Date;
  end: Date;
  titleKey?: string;
  timeOptions?: CountdownBoxOptions;
};

const defaultTimeOptions: CountdownBoxOptions = {
  seconds: true,
  minutes: true,
  hours: true,
  days: true,
};

const Countdown = ({
  start = new Date(),
  end,
  titleKey = "countdown-title",
  timeOptions = defaultTimeOptions,
}: CountdownProps) => {
  const { t } = useTranslation();
  const [startTime, setStartTime] = useState(start);
  const [endTime] = useState(end);

  const timeValues = useMemo(() => getTimeDiffObj(startTime, endTime), [startTime, endTime]);
  const qtyOfBoxes = useMemo(() => Object.values(timeOptions).filter((value) => value).length, [timeOptions]);

  useEffect(() => {
    if (startTime >= endTime) return;

    const interval = setInterval(() => {
      setStartTime((previousTime) => new Date(previousTime.getTime() + 1000));
    }, 1000);
    return () => clearInterval(interval);
  }, [endTime, startTime]);

  return (
    <Wrapper>
      <CountdownTitle>{t(titleKey)}</CountdownTitle>
      <BoxWrapper boxQty={qtyOfBoxes}>
        {timeOptions.days && <TimeBox time={timeValues.days} label={t("days")} />}
        {timeOptions.hours && <TimeBox time={timeValues.hours} label={t("hours")} />}
        {timeOptions.minutes && <TimeBox time={timeValues.minutes} label={t("minutes")} />}
        {timeOptions.seconds && <TimeBox time={timeValues.seconds} label={t("seconds")} />}
      </BoxWrapper>
    </Wrapper>
  );
};

export default Countdown;
