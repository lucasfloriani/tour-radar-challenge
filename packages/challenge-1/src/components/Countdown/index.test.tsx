import React from "react";
import { act, render, screen } from "@testing-library/react";
import { SECOND, MINUTE } from "../../helpers/time";
import { wrapper } from "../../tests/utils";

import Countdown from ".";

describe("Countdown component", () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it("should render correctly with default config", () => {
    const timeNow = new Date();
    const endTime = new Date(timeNow.getTime() + SECOND * 10);
    const { container } = render(<Countdown start={timeNow} end={endTime} />, { wrapper });
    expect(container).toMatchSnapshot();
  });

  it("should decrease the time", () => {
    const timeNow = new Date();
    const endTime = new Date(timeNow.getTime() + SECOND * 10);

    act(() => {
      render(<Countdown start={timeNow} end={endTime} />, { wrapper });
    });
    expect(screen.getAllByText("0").length).toBe(3);
    expect(screen.getAllByText("10").length).toBe(1);

    act(() => jest.advanceTimersByTime(SECOND));
    expect(screen.getAllByText("0").length).toBe(3);
    expect(screen.getAllByText("9").length).toBe(1);
  });

  it("should decrease the time to zero and stop", () => {
    const timeNow = new Date();
    const endTime = new Date(timeNow.getTime() + SECOND * 10);

    act(() => {
      render(<Countdown start={timeNow} end={endTime} />, { wrapper });
    });
    expect(screen.getAllByText("0").length).toBe(3);
    expect(screen.getAllByText("10").length).toBe(1);

    act(() => jest.advanceTimersByTime(SECOND * 10));
    expect(screen.getAllByText("0").length).toBe(4);

    act(() => jest.advanceTimersByTime(SECOND));
    expect(screen.getAllByText("0").length).toBe(4);
  });

  it("should decrease minutes", () => {
    const timeNow = new Date();
    const endTime = new Date(timeNow.getTime() + MINUTE);

    act(() => {
      render(<Countdown start={timeNow} end={endTime} />, { wrapper });
    });
    expect(screen.getAllByText("0").length).toBe(3);
    expect(screen.getAllByText("1").length).toBe(1);

    act(() => jest.advanceTimersByTime(SECOND));
    expect(screen.getAllByText("0").length).toBe(3);
    expect(screen.getAllByText("59").length).toBe(1);
  });
});
