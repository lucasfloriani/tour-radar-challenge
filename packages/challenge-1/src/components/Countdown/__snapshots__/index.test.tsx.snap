// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countdown component should render correctly with default config 1`] = `
.c1 {
  color: #000000;
  font-family: Helvetica,Arial,FreeSans,sans-serif;
  font-size: 0.8em;
  font-weight: 600;
  text-align: left;
  margin: 0;
}

.c5 {
  color: #000000;
  font-family: Helvetica,Arial,FreeSans,sans-serif;
  font-size: 2em;
  font-weight: bold;
  text-align: center;
  margin: 0;
}

.c6 {
  color: #000000;
  font-family: Helvetica,Arial,FreeSans,sans-serif;
  font-size: .6em;
  font-weight: normal;
  text-align: center;
  margin: 0;
}

.c4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 1px solid #000000;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 60px;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  padding-bottom: 2px;
  width: 60px;
}

.c7 {
  text-transform: uppercase;
}

.c0 {
  display: grid;
  gap: 3px;
}

.c3 {
  display: inline-grid;
  gap: 8px;
  grid-template-columns: repeat(4,auto);
}

.c2 {
  -webkit-letter-spacing: 3px;
  -moz-letter-spacing: 3px;
  -ms-letter-spacing: 3px;
  letter-spacing: 3px;
  text-transform: uppercase;
}

<div>
  <section
    class="c0"
  >
    <h4
      class="c1 c2"
      color="grayscale.100"
    >
      Starts in
    </h4>
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <p
          class="c5"
          color="grayscale.100"
        >
          0
        </p>
        <p
          class="c6 c7"
          color="grayscale.100"
        >
          days
        </p>
      </div>
      <div
        class="c4"
      >
        <p
          class="c5"
          color="grayscale.100"
        >
          0
        </p>
        <p
          class="c6 c7"
          color="grayscale.100"
        >
          hours
        </p>
      </div>
      <div
        class="c4"
      >
        <p
          class="c5"
          color="grayscale.100"
        >
          0
        </p>
        <p
          class="c6 c7"
          color="grayscale.100"
        >
          minutes
        </p>
      </div>
      <div
        class="c4"
      >
        <p
          class="c5"
          color="grayscale.100"
        >
          10
        </p>
        <p
          class="c6 c7"
          color="grayscale.100"
        >
          seconds
        </p>
      </div>
    </div>
  </section>
</div>
`;
